language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/BallerinaLanguageServer.java

```diff
@@ -131,36 +132,34 @@ public ExtendedLanguageClient getClient() {
         res.getCapabilities().setImplementationProvider(false);
 //        res.getCapabilities().setCodeLensProvider(new CodeLensOptions());
 
-        TextDocumentClientCapabilities textDocCapabilities = params.getCapabilities().getTextDocument();
-        ((BallerinaTextDocumentService) this.textService).setClientCapabilities(textDocCapabilities);
-
-        HashMap experimentalClientCapabilities = null;
+        HashMap expCapabilities = null;
         if (params.getCapabilities().getExperimental() != null) {
-            experimentalClientCapabilities =
-                    new Gson().fromJson(params.getCapabilities().getExperimental().toString(), HashMap.class);
+            expCapabilities = new Gson().fromJson(params.getCapabilities().getExperimental().toString(), HashMap.class);
         }
 
-        BallerinaWorkspaceService workspaceService = (BallerinaWorkspaceService) this.workspaceService;
-        workspaceService.setExperimentalClientCapabilities(experimentalClientCapabilities);
-
         // Set AST provider and examples provider capabilities
         HashMap<String, Object> experimentalServerCapabilities = new HashMap<>();
         experimentalServerCapabilities.put(AST_PROVIDER.getValue(), true);
         experimentalServerCapabilities.put(EXAMPLES_PROVIDER.getValue(), true);
         experimentalServerCapabilities.put(API_EDITOR_PROVIDER.getValue(), true);
-        if (experimentalClientCapabilities != null &&
-                experimentalClientCapabilities.get(INTROSPECTION.getValue()) != null) {
+        if (expCapabilities != null && expCapabilities.get(INTROSPECTION.getValue()) != null) {
             int port = ballerinaTraceListener.startListener();
             experimentalServerCapabilities.put(INTROSPECTION.getValue(), new ProviderOptions(port));
         }
-        if (experimentalClientCapabilities != null &&
-                experimentalClientCapabilities.get(SEMANTIC_SYNTAX_HIGHLIGHTER.getValue()) != null) {
+        if (expCapabilities != null &&
+                expCapabilities.get(SEMANTIC_SYNTAX_HIGHLIGHTER.getValue()) != null) {
             experimentalServerCapabilities.put(SEMANTIC_SYNTAX_HIGHLIGHTER.getValue(), true);
             String[][] scopes = getScopes();
             experimentalServerCapabilities.put(SEMANTIC_SCOPES.getValue(), scopes);
         }
         res.getCapabilities().setExperimental(experimentalServerCapabilities);
 
+
+        TextDocumentClientCapabilities txtCapabilities = params.getCapabilities().getTextDocument();
+        WorkspaceClientCapabilities wsCapabilities = params.getCapabilities().getWorkspace();
+        LSClientCapabilitiesHolder capabilities = LSClientCapabilitiesHolder.getInstance();
```

Comment:
Let's deviate from the singleton and instead try to pass the capabilities to the services?